pkg/build/buildkit/wait.go

```diff
@@ -0,0 +1,135 @@
+
+	"github.com/moby/buildkit/client"
+	"github.com/okteto/okteto/pkg/env"
+	"github.com/okteto/okteto/pkg/log/io"
+)
+
+const (
+	// maxWaitTime is the maximum time to wait for buildkit to be available
+	maxWaitTime = 10 * time.Minute
+
+	// retryTime is the time to wait between retries
+	retryTime = 5 * time.Second
+
+	// maxBuildkitWaitTimeEnvVar is the environment variable to set the max wait time for buildkit
+	maxBuildkitWaitTimeEnvVar = "OKTETO_BUILDKIT_WAIT_TIMEOUT"
+
+	// retryBuildkitTimeEnvVar is the environment variable to set the retry time for buildkit
+	retryBuildkitIntervalEnvVar = "OKTETO_RETRY_BUILDKIT_INTERVAL"
+)
+
+// sleeper defines an interface for sleeping
+type sleeper interface {
+	Sleep(duration time.Duration)
+}
+
+// RealSleeper implements the Sleeper interface using time.Sleep
+type RealSleeper struct{}
+
+// Sleep sleeps for the specified duration
+func (s *RealSleeper) Sleep(duration time.Duration) {
+	time.Sleep(duration)
+}
+
+type buildkitClientFactoryWaiter interface {
+	GetBuildkitClient(ctx context.Context) (clientInfoRetriever, error)
+}
+
+type clientInfoRetriever interface {
+	Info(ctx context.Context) (*client.Info, error)
+}
+
+// Waiter encapsulates the logic to check if the BuildKit server is up and running
+type Waiter struct {
+	sleeper               sleeper
+	buildkitClientFactory buildkitClientFactoryWaiter
+	logger                *io.Controller
+	waitingTime           time.Duration
+	maxWaitTime           time.Duration
+	retryInterval         time.Duration
+}
+
+type buildkitClientFactoryToWait struct {
+	factory *ClientFactory
+}
+
+func (b *buildkitClientFactoryToWait) GetBuildkitClient(ctx context.Context) (clientInfoRetriever, error) {
+	return b.factory.GetBuildkitClient(ctx)
+}
+
+func buildkitClientFactoryToWaitFactory(factory *ClientFactory) buildkitClientFactoryWaiter {
+	return &buildkitClientFactoryToWait{factory}
+}
+
+// NewBuildkitClientWaiter creates a new buildkitWaiter
+func NewBuildkitClientWaiter(factory *ClientFactory, logger *io.Controller) *Waiter {
+	return &Waiter{
+		maxWaitTime:           env.LoadTimeOrDefault(maxBuildkitWaitTimeEnvVar, maxWaitTime),
+		retryInterval:         env.LoadTimeOrDefault(retryBuildkitIntervalEnvVar, retryTime),
+		sleeper:               &RealSleeper{},
+		buildkitClientFactory: buildkitClientFactoryToWaitFactory(factory),
+		logger:                logger,
+	}
+}
+
+func (bw *Waiter) GetWaitingTime() time.Duration {
+	return bw.waitingTime
+}
+
+// WaitUntilIsUp waits for the BuildKit server to become available
+func (bw *Waiter) WaitUntilIsUp(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, bw.maxWaitTime)
+	defer cancel()
+
+	sp := bw.logger.Out().Spinner("Waiting for BuildKit service to be ready...")
+	sp.Start()
+	defer sp.Stop()
+
+	startWaitingTime := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			bw.waitingTime = bw.maxWaitTime
+			return fmt.Errorf("buildkit service not available for %v", bw.maxWaitTime)
```

Comment:
nit: the following scenarios can be a _misconfiguration_ but  when `bw.retryInterval` is grather than `bw.maxWaitTime`, the total time should be `bw.retryInterval`.

Example:

```
OKTETO_RETRY_BUILDKIT_INTERVAL=20s OKTETO_BUILDKIT_WAIT_TIMEOUT=1s okd build
 x  Error building service 'frontend': buildkit service not available for 1s
```

Reply:
I agree but it's weird that the max waiting time that a user has set is lower than the error we show